README.md
# Omni Chat Backend

_Trusted, Extensible, Better Chat_

![Cover](branding/facebook_cover_photo_2.png)

Read about the service [here](docs/about.md).

To view a specific version's docs, go to `https://github.com/neelkamath/omni-chat-backend/tree/<VERSION>`, where `<VERSION>` is the [release tag](https://github.com/neelkamath/omni-chat-backend/tags) (e.g., `v0.1.1`).

Here are the guides for running the server using [Docker Compose](docs/docker-compose.md) (recommended for local development), and the [cloud](docs/cloud.md) (recommended for production).

Read these [recommendations](docs/frontend-recommendations.md) if you're a developer creating a frontend UI which uses this API.

Here are some notable projects built using the Omni Chat API. If you've built an app, library, etc., create a [GitHub issue](https://github.com/neelkamath/omni-chat-backend/issues/new) to have it added to this list.
- (Official) [Web app](https://github.com/neelkamath/omni-chat-web)
- (Official) [JavaScript API wrapper library](https://www.npmjs.com/package/@neelkamath/omni-chat)

## Usage

**Note**: Operations starting with `Query`, `Mutation`, and `Subscription` (e.g., `Query.readChat`) are from the GraphQL API. Operations starting with `/` (e.g., `/profile-image`) are from the REST API.

- [Docs](docs/api.md)
    - Download the `rest-api.html` asset from a [release](https://github.com/neelkamath/omni-chat-backend/releases). It'll be referenced in the docs.
    - Optionally, generate a wrapper for the GraphQL API using [GraphQL Code Generator](https://graphql-code-generator.com/) on [`schema.graphqls`](src/main/resources/schema.graphqls).
    - Optionally, generate a wrapper for the REST API using [OpenAPI Generator](https://openapi-generator.tech/) on [`openapi.yaml`](docs/openapi.yaml). Note that backwards compatible REST API updates don't guarantee backwards compatible wrappers. For example, a wrapper for REST API v0.3.1 may not be backwards compatible with a wrapper for REST API v0.3.0.
- [Changelog](docs/CHANGELOG.md)
- [Branding assets](branding)

## [Contributing](docs/CONTRIBUTING.md)

## Credits

[`dockerize`](docker/dockerize) was taken from [jwilder](https://github.com/jwilder/dockerize).

## License

This project is under the [MIT License](LICENSE).


docs/CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

The entire project (i.e., the GraphQL API, REST API, and server) uses the same version number. Major and minor versions get based off of the API (i.e., the GraphQL and REST APIs). For example, if the server has a backward incompatible change such as a DB schema update, but the APIs haven't changed, then only the patch number gets bumped. Another example is if the GraphQL API hasn't changed, but the format of the HTTP request used to send the GraphQL document has changed, then the major version gets bumped.

## [0.25.0](https://github.com/neelkamath/omni-chat-backend/releases/tag/v0.25.0) - 2021-07-17

### Fixed

- Fix the production `docker-compose.yml`.
- Fix the documentation for running the application in production.
- Fix the REST API docs.
- Fix `Mutation.makeGroupChatAdmins`.

## [0.24.0](https://github.com/neelkamath/omni-chat-backend/releases/tag/v0.24.0) - 2021-07-15

### Changed

- Rename the `pic-type` parameter to `image-type` in HTTP GET `/image-message`, HTTP GET `/profile-image`, and HTTP GET `/group-chat-image`.
- Rename `/pic-message` to `/image-message`.
- Rename `/profile-pic` to `/profile-image`.
- Rename `/group-chat-pic` to `/group-chat-image`.
- Rename `Mutation.deleteProfilePic` to `Mutation.deleteProfileImage`.
- Rename `Mutation.deleteGroupChatPic` to `Mutation.deleteGroupChatImage`.
- Rename `type UpdatedProfilePic` to `type UpdatedProfileImage`.
- Rename `type UpdatedGroupChatPic` to `type UpdatedGroupChatImage`.
- Rename `type PicMessage` to `type ImageMessage`.
- Rename `type BookmarkedPicMessage` to `type BookmarkedImageMessage`.
- Rename `type NewPicMessage` to `type NewImageMessage`.
- Remove authentication for HTTP GET `/group-chat-image`.

### Fixed

- Fix media downloads by sending back the send back the `Content-Disposition` header with the filename so that the browser downloads the file with the correct media type. The following operations have been fixed:
    - HTTP GET `/profile-image`
    - HTTP GET `/group-chat-image`
    - HTTP GET `/image-message`
    - HTTP GET `/audio-message`
    - HTTP GET `/video-message`
    - HTTP GET `/doc-message`

## [0.23.0](https://github.com/neelkamath/omni-chat-backend/releases/tag/v0.23.0) - 2021-07-04

### Changed

- Improve error handling for HTTP POST `/pic-message`, HTTP POST `/audio-message`, HTTP POST `/video-message`, and HTTP POST `/doc-message`.

    Previously, creating a message in a broadcast chat would cause an HTTP status code of 401 to be returned. Now, an HTTP status code of 400 is returned with the following `application/json` response body:

    ```json
    {
      "reason": "MUST_BE_ADMIN"
    }
    ```
- Add the field `chatId` to `interface BareGroupChat` and `type GroupChatInfo`.
- Rename `Query.readStars` to `Query.readBookmarks`.
- Rename `Mutation.star` to `Mutation.createBookmark`.
- Rename `Mutation.unstar` to `Mutation.deleteBookmark`.
- Rename `type UnstarredChat` to `type UnbookmarkedChat`.
- Rename `type StarredMessagesConnection` to `type BookmarkedMessagesConnection`.
- Rename `type StarredMessageEdge` to `type BookmarkedMessageEdge`.
- Rename the `hasStar` field to `isBookmarked` for the following:
    - `interface Message`
    - `type TextMessage`
    - `type ActionMessage`
    - `type PicMessage`
    - `type PollMessage`
    - `type AudioMessage`
    - `type GroupChatInviteMessage`
    - `type DocMessage`
    - `type VideoMessage`
    - `type UpdatedMessage`
- Rename `interface StarredMessage` to `interface BookmarkedMessage`.
- Rename `type StarredTextMessage` to `type BookmarkedTextMessage`.
- Rename `type StarredActionMessage` to `type BookmarkedActionMessage`.
- Rename `type StarredPicMessage` to `type BookmarkedPicMessage`.
- Rename `type StarredPollMessage` to `type BookmarkedPollMessage`.
- Rename `type StarredAudioMessage` to `type BookmarkedAudioMessage`.
- Rename `type StarredGroupChatInviteMessage` to `type BookmarkedGroupChatInviteMessage`.
- Rename `type StarredDocMessage` to `type BookmarkedDocMessage`.
- Rename `type StarredVideoMessage` to `type BookmarkedVideoMessage`.
- Change the size for media uploads and downloads from 5 MiB to 3 MiB for the following:
    - `/profile-pic`
    - `/group-chat-pic`
    - `/pic-message`
    - `/audio-message`
    - `/video-message`
    - `/doc-message`
- Allow creating an invitation for the chat itself using `Mutation.createGroupChatInviteMessage` instead of returning a `type InvalidChatId`.
- Allow forwarding a message in the chat it's from using `Mutation.forwardMessage` instead of returning a `type InvalidChatId`.
- Allow forwarding a group chat invite in the chat it's for using `Mutation.forwardMessage` instead of returning a `type InvalidChatId`.

### Fixed

- Fix `Mutation.leaveGroupChat` and `Mutation.removeGroupChatUsers`. They used to crash if every user left the chat when there were group chat invitations for the chat in other chats.
- Fix `Subscription.subscribeToMessages` not sending back `type UnbookmarkedChat`s.

## [0.22.0](https://github.com/neelkamath/omni-chat-backend/releases/tag/v0.22.0) - 2021-06-29

### Added

- `Query.readAllowedEmailAddressDomains`
- `union ReadAccountResult`

### Changed

- Change `type UpdatedPollMessage` to retrieve the entire poll instead for ease of use.
- Return a `union ReadAccountResult` instead of a `type Account` from `Query.readAccount`.
- Remove the `lastOnline` field from `type OnlineStatus`.
- Remove the `state` and `statuses` fields from the following:
    - `interface Message`
    - `type TextMessage`
    - `type ActionMessage`
    - `type PicMessage`
    - `type PollMessage`
    - `type AudioMessage`
    - `type GroupChatInviteMessage`
    - `type DocMessage`
    - `type VideoMessage`
    - `interface StarredMessage`

src/main/kotlin/Auth.kt
package com.neelkamath.omniChat

import com.neelkamath.omniChat.db.deleteUserFromDb
import com.neelkamath.omniChat.db.negotiateUserUpdate
import com.neelkamath.omniChat.db.tables.Users
import com.neelkamath.omniChat.graphql.routing.*
import io.ktor.http.*
import org.keycloak.admin.client.Keycloak
import org.keycloak.admin.client.KeycloakBuilder
import org.keycloak.admin.client.resource.RealmResource
import org.keycloak.admin.client.resource.UsersResource
import org.keycloak.authorization.client.AuthzClient
import org.keycloak.authorization.client.Configuration
import org.keycloak.authorization.client.util.HttpResponseException
import org.keycloak.representations.idm.ClientRepresentation
import org.keycloak.representations.idm.CredentialRepresentation
import org.keycloak.representations.idm.RealmRepresentation
import org.keycloak.representations.idm.UserRepresentation

private const val REALM_NAME = "omni-chat"
private const val CLIENT_ID = "server"
private val config: Configuration = Configuration().apply {
    realm = REALM_NAME
    authServerUrl = "${System.getenv("KEYCLOAK_URL")}/auth"
    sslRequired = "external"
    resource = CLIENT_ID
    credentials = mapOf("secret" to System.getenv("KEYCLOAK_CLIENT_SECRET"))
    confidentialPort = 0
}
private val keycloak: Keycloak = KeycloakBuilder
    .builder()
    .serverUrl("${System.getenv("KEYCLOAK_URL")}/auth")
    .realm("master")
    .username(System.getenv("KEYCLOAK_USER"))
    .password(System.getenv("KEYCLOAK_PASSWORD"))
    .clientId("admin-cli")
    .build()
private lateinit var realm: RealmResource
private val omniChatRealm: Lazy<RealmRepresentation> = lazy {
    RealmRepresentation().apply {
        realm = REALM_NAME
        displayName = "Omni Chat"
        isEnabled = true
        isVerifyEmail = true
        isEditUsernameAllowed = true
        smtpServer = mapOf(
            "host" to System.getenv("KEYCLOAK_SMTP_HOST"),
            "port" to System.getenv("KEYCLOAK_SMTP_TLS_PORT"),
            "from" to System.getenv("KEYCLOAK_SMTP_USER"),
            "starttls" to "true",
            "auth" to "true",
            "user" to System.getenv("KEYCLOAK_SMTP_USER"),
            "password" to System.getenv("KEYCLOAK_SMTP_PASSWORD")
        )
    }
}

fun UserRepresentation.toAccount(): Account {
    val user = Users.read(id)
    return Account(user.id, Username(username), email, firstName, lastName, user.bio)
}

/**
 * Sets up account management.
 *
 * This takes a few seconds to run if the auth system hasn't been set up, and must be run before any account-related
 * activities are performed.
 */
fun setUpAuth() {
    val shouldBuild = REALM_NAME !in keycloak.realms().findAll().map { it.realm }
    if (shouldBuild) keycloak.realms().create(omniChatRealm.value)
    realm = keycloak.realm(REALM_NAME)
    if (shouldBuild) createClient()
}

/** Creates the client used to access the [realm] programmatically. Don't call this if the client exists.  */
private fun createClient() {
    realm.clients().create(
        ClientRepresentation().apply {
            clientId = CLIENT_ID
            secret = System.getenv("KEYCLOAK_CLIENT_SECRET")
            isDirectAccessGrantsEnabled = true
            rootUrl = System.getenv("KEYCLOAK_URL")
        }
    )
}

/** Whether the [login] is valid. */
fun isValidLogin(login: Login): Boolean = try {
    AuthzClient.create(config).obtainAccessToken(login.username.value, login.password.value)
    true
} catch (exception: HttpResponseException) {
    /*
    Keycloak returns an HTTP status code of 400 (Bad Request) or 401 (Unauthorized) if the login is invalid. It's
    possible that a Bad Request is returned due to a bug on Omni Chat's server (i.e., the login is valid). However,
    were this to occur, every request would be invalid, and it would be obvious that there's a bug.
     */
    if (exception.statusCode in listOf(HttpStatusCode.Unauthorized.value, HttpStatusCode.BadRequest.value)) false
    else throw exception
}

fun emailAddressExists(email: String): Boolean = email in realm.users().list().map { it.email }

/**
 * Creates a new account in the auth system, saves it to the DB via [Users.create], and sends the user a verification
 * email.
 */
fun createUser(account: AccountInput) {
    realm.users().create(account.toUserRepresentation())
    val userId = realm
        .users()
        .search(account.username.value)
        .first { it.username == account.username.value }
        .id
    Users.create(userId, account.bio)
    sendEmailAddressVerification(account.emailAddress)
}

/** Sends an email to the user to verify their email [address]. */
fun sendEmailAddressVerification(address: String) {
    val userId = readUserByEmailAddress(address).id.let(Users::read).uuid.toString()
    realm.users().get(userId).sendVerifyEmail()
}

/** Sends an email for the user to reset their password. */
fun resetPassword(email: String) {
    val userId = readUserByEmailAddress(email).id
    realm.users().get(Users.read(userId).uuid.toString()).executeActionsEmail(listOf("UPDATE_PASSWORD"))
}

private fun AccountInput.toUserRepresentation(): UserRepresentation = UserRepresentation().also {
    it.username = username.value
    it.credentials = createCredentials(password)
    it.email = emailAddress
    it.firstName = firstName
    it.lastName = lastName
    it.isEnabled = true
}

fun readUserByUsername(username: Username): Account =
    realm.users().search(username.value).first { it.username == username.value }.toAccount()

fun isEmailVerified(userId: Int): Boolean = readUser(userId).isEmailVerified

fun readUserById(userId: Int): Account = readUser(userId).toAccount()

private fun readUser(userId: Int): UserRepresentation = realm.users().list().first { Users.read(it.id).id == userId }

private fun readUserByEmailAddress(email: String): Account =
    realm.users().list().first { it.email == email }.toAccount()

/**
 * Case-insensitively [query]s every user's username, first name, last name, and email address.
 *
 * @see [Users.search]
 */
fun searchUsers(query: String): List<Account> = with(realm.users()) {
    searchBy(username = query) +
            searchBy(firstName = query) +
            searchBy(lastName = query) +
            searchBy(emailAddress = query)
}.distinctBy { it.id }.map { it.toAccount() }

/** Convenience function for [UsersResource.search] which provides only the relevant parameters with names. */
private fun UsersResource.searchBy(
    username: String? = null,
    firstName: String? = null,
    lastName: String? = null,
    emailAddress: String? = null
): List<UserRepresentation> = search(username, firstName, lastName, emailAddress, null, null)

/** Calls [negotiateUserUpdate]. */
fun updateUser(id: Int, update: AccountUpdate) {
    val user = readUser(id)
    if (update.emailAddress != null && user.email != update.emailAddress) user.isEmailVerified = false
    user.update(update)
    realm.users().get(Users.read(id).uuid.toString()).update(user)
    Users.updateBio(id, update.bio)
    negotiateUserUpdate(id)
}

fun isUsernameTaken(username: Username): Boolean {
    val results = realm.users().search(username.value)
    return results.isNotEmpty() && results.any { it.username == username.value }
}

/** Deletes the user [id] from the auth system after calling [deleteUserFromDb]. */
fun deleteUser(id: Int) {
    realm.users().delete(Users.read(id).uuid.toString())
    deleteUserFromDb(id)
}

/** Applies this [update]. */
private fun UserRepresentation.update(update: AccountUpdate) {
    update.username?.let { username = it.value }
    update.password?.let { credentials = createCredentials(it) }
    update.emailAddress?.let { email = it }
    update.firstName?.let { firstName = it }
    update.lastName?.let { lastName = it }
}

private fun createCredentials(password: Password): List<CredentialRepresentation> = listOf(
    CredentialRepresentation().apply {
        type = CredentialRepresentation.PASSWORD
        value = password.value
    }
)

/** Whether the [emailAddress]'s domain (e.g., `"example.com"`) is allowed by this Omni Chat instance. */
fun hasAllowedDomain(emailAddress: String): Boolean {
    val domains = System.getenv("ALLOWED_DOMAINS") ?: return true
    if (domains.isEmpty()) return true
    return emailAddress.substringAfter("@") in domains.split(",")
}